funcbench/bench.go

```diff
@@ -0,0 +1,142 @@
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+// TODO: Add unit test.
+type Benchmarker struct {
+	logger Logger
+
+	benchFunc      string
+	benchTime      time.Duration
+	benchTimeout   time.Duration
+	resultCacheDir string
+
+	c    *commander
+	repo *git.Repository
+}
+
+func newBenchmarker(logger Logger, env Environment, c *commander, benchTime time.Duration, benchTimeout time.Duration, resultCacheDir string) *Benchmarker {
+	return &Benchmarker{
+		logger:         logger,
+		benchFunc:      env.BenchFunc(),
+		benchTime:      benchTime,
+		benchTimeout:   benchTimeout,
+		c:              c,
+		repo:           env.Repo(),
+		resultCacheDir: resultCacheDir,
+	}
+}
+
+func (b *Benchmarker) benchOutFileName(commit plumbing.Hash) (string, error) {
+	// Sanitize bench func.
+	bb := bytes.Buffer{}
+	e := base64.NewEncoder(base64.StdEncoding, &bb)
+	if _, err := e.Write([]byte(b.benchFunc)); err != nil {
+		return "", err
+	}
+	if err := e.Close(); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s-%s.out", bb.String(), commit.String()), nil
+}
+
+func (b *Benchmarker) execBenchmark(pkgRoot string, commit plumbing.Hash) (out string, err error) {
+	fileName, err := b.benchOutFileName(commit)
+	if err != nil {
+		return "", err
+	}
+
+	if _, err := ioutil.ReadFile(filepath.Join(b.resultCacheDir, fileName)); err == nil {
+		fmt.Println("Found previous results for ", fileName, b.benchFunc, "Reusing.")
+		return filepath.Join(b.resultCacheDir, fileName), nil
+	}
+
+	// TODO(bwplotka): Allow memprofiles.
+	extraArgs := []string{"-benchtime", b.benchTime.String()}
+	extraArgs = append(extraArgs, "-timeout", b.benchTimeout.String())
+	benchCmd := []string{"bash", "-c", strings.Join(append(append([]string{"cd", pkgRoot, "&&", "go", "test", "-run", "^$", "-bench", fmt.Sprintf("^%s$", b.benchFunc), "-benchmem"}, extraArgs...), "./..."), " ")}
+
+	b.logger.Println("Executing benchmark cmd:", benchCmd)
+	out, err = b.c.exec(benchCmd...)
+	if err != nil {
+		return "", errors.Wrap(err, "benchmark ended with an error.")
+	}
+
+	fn := fileName
+	if b.resultCacheDir != "" {
+		fn = filepath.Join(b.resultCacheDir, fileName)
+	}
+	if err := ioutil.WriteFile(fn, []byte(out), os.ModePerm); err != nil {
+		return "", err
+	}
+	return fn, nil
+}
+
+func parseFile(path string) (parse.Set, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	bb, err := parse.ParseSet(f)
+	if err != nil {
+		return nil, err
+	}
+	return bb, nil
+}
+
+func (b *Benchmarker) compareBenchmarks(beforeFile, afterFile string) ([]BenchCmp, error) {
+	before, err := parseFile(beforeFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "open %s", beforeFile)
+	}
+
+	after, err := parseFile(afterFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "open %s", afterFile)
+	}
+
+	// Use benchcmp library directly.
+	// TODO(bwplotka): benchstat is new thing - we might add choice for funcbench to choose from? (:
```

Comment:
Surely `benchcmp` will be dropped. I prefer `benchstat`.

Reply:
I would not be that sure, let's decide later.